cmake/kokkos_arch.cmake

```diff
@@ -574,15 +574,37 @@ CHECK_AMDGPU_ARCH(VEGA906 gfx906) # Radeon Instinct MI50 and MI60
 CHECK_AMDGPU_ARCH(VEGA908 gfx908) # Radeon Instinct MI100
 CHECK_AMDGPU_ARCH(VEGA90A gfx90a) # Radeon Instinct MI200
 
+FUNCTION(SET_AND_CHECK_AMD_ARCH ARCH FLAG)
+  SET(AMD_ARCHITECTURE ${ARCH} PARENT_SCOPE)
+  KOKKOS_SET_OPTION(ARCH_${ARCH} ON)
+  CHECK_AMDGPU_ARCH(${ARCH} ${FLAG})
+ENDFUNCTION()
+
 IF(KOKKOS_ENABLE_HIP AND NOT AMDGPU_ARCH_ALREADY_SPECIFIED)
   IF(KOKKOS_CXX_COMPILER_ID STREQUAL HIPCC)
     FIND_PROGRAM(ROCM_ENUMERATOR rocm_agent_enumerator)
     EXECUTE_PROCESS(COMMAND ${ROCM_ENUMERATOR} OUTPUT_VARIABLE GPU_ARCHS)
     STRING(LENGTH "${GPU_ARCHS}" len_str)
     # enumerator always output gfx000 as the first line
     IF(${len_str} LESS 8)
-      MESSAGE(SEND_ERROR "HIP enabled but no AMD GPU architecture currently enabled. "
-                         "Please enable one AMD GPU architecture via -DKokkos_ARCH_{..}=ON'.")
+      MESSAGE(SEND_ERROR "HIP enabled but no AMD GPU architecture could be automatically detected. "
+                         "Please manually enable one AMD GPU architecture via -DKokkos_ARCH_{..}=ON'.")
+    # check for known gpu archs, otherwise error out
+    ELSE()
+      STRING(REGEX MATCH "(gfx900|gfx906|gfx908|gfx90a)" DETECTED_GPU_ARCH ${GPU_ARCHS})
+      IF("${DETECTED_GPU_ARCH}" STREQUAL "gfx900")
+        SET_AND_CHECK_AMD_ARCH(VEGA900 ${DETECTED_GPU_ARCH})
+      ELSEIF("${DETECTED_GPU_ARCH}" STREQUAL "gfx906")
+        SET_AND_CHECK_AMD_ARCH(VEGA906 ${DETECTED_GPU_ARCH})
+      ELSEIF("${DETECTED_GPU_ARCH}" STREQUAL "gfx908")
+        SET_AND_CHECK_AMD_ARCH(VEGA908 ${DETECTED_GPU_ARCH})
+      ELSEIF("${DETECTED_GPU_ARCH}" STREQUAL "gfx90a")
+        SET_AND_CHECK_AMD_ARCH(VEGA90a ${DETECTED_GPU_ARCH})
+      ELSE()
+        MESSAGE(SEND_ERROR "HIP enabled but no AMD GPU architecture could be automatically detected. "
+                         "Please manually enable one AMD GPU architecture via -DKokkos_ARCH_{..}=ON'.")
+      ENDIF()
+      LIST(APPEND KOKKOS_ENABLED_ARCH_LIST ${AMD_ARCHITECTURE})
```

Comment:
Move this to `SET_AND_CHECK_AMD_ARCH` to avoid passing the extra variable.

Reply:
done in [54bca5](https://github.com/kokkos/kokkos/pull/5327/commits/54bca53977483dc918c4bc731a6042e9b94d0bab)